### PR TITLE
Copter: fix short waypoint duration

### PR DESCRIPTION
**Pull request by lthall**
This PR does two things.
1. It fixes the problem where very short waypoints, near zero length, have a minimum duration of 8 times tj. To fix this I have moved to a constant Snap formulation where instead of dropping just the Jp we drop both Jp and tj. This means that as the length tends towards zero, tj tends towards zero.
2. It optimises the corner timing to achieve maximum velocity though a corner in all cases.
image

Master
image

PR
image

### CODE REVIEW

File: libraries/AC_WPNav/AC_WPNav.cpp [hunk 1]
```diff
@@ -868,14 +869,14 @@ void AC_WPNav::calc_scurve_jerk_and_jerk_time()
...
+    // calculate maximum snap
+    // Snap (the rate of change of jerk) uses the attitude control input time constant because multicopters
     // lean to accelerate. This means the change in angle is equivalent to the change in acceleration
-    const float jounce = MIN(_attitude_control.get_accel_roll_max_radss() * GRAVITY_MSS, _attitude_control.get_accel_pitch_max_radss() * GRAVITY_MSS);
-    if (is_positive(jounce)) {
-        _scurve_jerk_time = MAX(_attitude_control.get_input_tc(), 0.5f * _scurve_jerk * M_PI / jounce);
-    } else {
-        _scurve_jerk_time = MAX(_attitude_control.get_input_tc(), 0.1f);
+    _scurve_snap = (_scurve_jerk * M_PI) / (2.0 * MAX(_attitude_control.get_input_tc(), 0.1f));
+    const float snap = MIN(_attitude_control.get_accel_roll_max_radss() * GRAVITY_MSS, _attitude_control.get_accel_pitch_max_radss() * GRAVITY_MSS);
```
Comment (IamPete1) [inline libraries/AC_WPNav/AC_WPNav.cpp:876]:
```suggestion
    const float snap = MIN(_attitude_control.get_accel_roll_max_radss(), _attitude_control.get_accel_pitch_max_radss()) * GRAVITY_MSS;
```

Small code efficiency improvement.

File: libraries/AP_Math/SCurve.cpp [hunk 2]
```diff
@@ -597,7 +623,7 @@ void SCurve::advance_time(float dt)
 // calculate the jerk, acceleration, velocity and position at the provided time
 void SCurve::get_jerk_accel_vel_pos_at_time(float time_now, float &Jt_out, float &At_out, float &Vt_out, float &Pt_out) const
 {
-    if ((num_segs != segments_max) || !is_positive(jerk_time)) {
+    if ((num_segs != segments_max)) {
```
Comment (IamPete1) [inline libraries/AP_Math/SCurve.cpp:626]:
```suggestion
    if (num_segs != segments_max) {
```

**Review by IamPete1 — COMMENTED** [review]
Code looks good, I don't see any potential divide by zeros or anything of that nature. Maths seems logical, but I do not understand the Scurve equations well enough to spot any issues. 

I think we just need lots of testing.

**Comment by lthall** [comment]
For those reviewing this. The main risk is that I don't handle a zero tj somwhere. Previously it was always set to a non-zero number, now it can be zero.

File: libraries/AP_Math/SCurve.cpp [hunk 3]
```diff
@@ -264,27 +266,30 @@ void SCurve::set_speed_max(float speed_xy, float speed_up, float speed_down)
     if (!is_equal(vel_max, segment[SEG_ACCEL_END].end_vel)) {
         // add velocity adjustment
         // check there is enough time to make velocity change
+
```
Comment (rmackay9) [inline libraries/AP_Math/SCurve.cpp:269]: got a little white space issue here.  might as well remove this line.

File: libraries/AP_Math/SCurve.cpp [hunk 4]
```diff
@@ -588,6 +596,24 @@ bool SCurve::braking() const
     return time >= segment[SEG_CONST].end_time;
 }
 
+// return true if the sequence is braking to a stop
```
Comment (rmackay9) [inline libraries/AP_Math/SCurve.cpp:599]: little copy-paste error on the comment I think.

File: libraries/AP_Math/SCurve.cpp [hunk 5]
```diff
@@ -588,6 +596,24 @@ bool SCurve::braking() const
...
+float SCurve::time_turn_in() const
+{
+    if (num_segs != segments_max) {
+        return 0.0;
+    }
+    return segment[SEG_TURN_IN].end_time;
+}
+
+// return true if the sequence is braking to a stop
```
Comment (rmackay9) [inline libraries/AP_Math/SCurve.cpp:608]: copy pasted comment here too..

File: libraries/AP_Math/SCurve.cpp [hunk 6]
```diff
@@ -760,9 +804,24 @@ void SCurve::calculate_path(float tj, float Jm, float V0, float Am, float Vm, fl
...
 
-    Am = MIN(MIN(Am, (Vm - V0) / (2.0f * tj)), (L + 4.0f * V0 * tj) / (4.0f * sq(tj)));
-    if (fabsf(Am) < Jm * tj) {
-        Jm = Am / tj;
+    float tj = Jm * M_PI / (2 * Sm);
+    float At = MIN(MIN(Am, 
+        (Vm - V0) / (2.0f * tj) ), 
+        (L + 4.0f * V0 * tj) / (4.0f * sq(tj)) );
+    if (fabsf(At) < Jm * tj) {
+        if (is_zero(V0)){
```
Comment (rmackay9) [inline libraries/AP_Math/SCurve.cpp:812]: nitpick: add a space before the curly bracket..

File: libraries/AP_Math/SCurve.h [hunk 7]
```diff
@@ -127,6 +127,12 @@ class SCurve {
     // return true if the sequence is braking to a stop
     bool braking() const WARN_IF_UNUSED;
 
+    // return true if the sequence is braking to a stop
```
Comment (rmackay9) [inline libraries/AP_Math/SCurve.h:130]: here too we should update the comment above the new methods.

File: libraries/AP_Math/SCurve.h [hunk 8]
```diff
@@ -53,21 +53,21 @@ class SCurve {
     void init();
 
     // calculate the segment times for the trigonometric S-Curve path defined by:
-    // tj - duration of the raised cosine jerk profile (aka jerk time)
-    // Jm - maximum value of the raised cosine jerk profile (aka jerk max)
+    // Sm - duration of the raised cosine jerk profile
```
Comment (rmackay9) [inline libraries/AP_Math/SCurve.h:56]: the comment next to "Sm" should be, "maximum snap magnitude" perhaps?

File: libraries/AP_Math/SCurve.h [hunk 9]
```diff
@@ -87,7 +87,7 @@ class SCurve {
     // target_pos should be set to this segment's origin and it will be updated to the current position target
     // target_vel and target_accel are updated with new targets
     // returns true if vehicle has passed the apex of the corner
-    bool advance_target_along_track(SCurve &prev_leg, SCurve &next_leg, float wp_radius, bool fast_waypoint, float dt, Vector3f &target_pos, Vector3f &target_vel, Vector3f &target_accel) WARN_IF_UNUSED;
+    bool advance_target_along_track(SCurve &prev_leg, SCurve &next_leg, float wp_radius, float accel_corner, bool fast_waypoint, float dt, Vector3f &target_pos, Vector3f &target_vel, Vector3f &target_accel) WARN_IF_UNUSED;
```
Comment (rmackay9) [inline libraries/AP_Math/SCurve.h:90]: It looks like we're missing descriptions of some of the arguments in the comments above this method.

- next_leg
- accel_corner

**Comment by rmackay9** [comment]
Merged, thanks!

**Comment by muramura** [comment]
This PR is generating a lot of WP shortcuts.
I think we lose the benefit of the S-curve with this PR.

**Comment by lthall** [comment]
What do you mean?
Do you have a log?

**Pull request merged**